plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,539 @@
+
+        // auto redeploy all models of this local node
+        threadPool.executor(DEPLOY_THREAD_POOL).execute(() -> {
+            long sleepTimeUnit = 1000L;
+            int numRetries = 4;
+            int numTries = 0;
+
+            // wait the index ".plugins-ml-model" finish initialization and get all the models to
+            // be deployed
+            while (numTries < numRetries) {
+                ++numTries;
+                log.info("trying to auto-redeploy model. try {} of {}", numTries, numRetries);
+
+                try {
+                    TimeUnit.MILLISECONDS.sleep(sleepTimeUnit * numTries);
+
+                    autoReDeployModel(localNodeId);
+                    break;
+                } catch (InterruptedException | RuntimeException e) {
+                    log.error("exception while auto-redeploying model.", e);
+                }
+            }
+        });
+    }
+
+    /**
+     * auto re-deploy model under the local node id<br>
+     *
+     * @param localNodeId local node id
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoReDeployModel(String localNodeId) throws MLException {
+        isMasterNode = RestActionUtils.isMasterNode(clusterService, localNodeId);
+        if (!isMasterNode && hasRoleName(ML_ROLE_NAME) && hasRoleName(DATA_ROLE_NAME)) {
+            // if the local node are both ml and data node,it should throw MLException
+            throw new MLException("the node " + localNodeId + " shouldn't be both ml and data node");
+        }
+
+        StepListener<SearchResponse> queryModelStep = new StepListener<>();
+
+        queryModels(ActionListener.wrap(queryModelStep::onResponse, queryModelStep::onFailure));
+
+        // At opensearch startup,to see the config item "plugins.ml_commons.only_run_on_ml_node" is true
+        // or false
+        log.info("only run on ml node? {} ", onlyRunOnMlNode);
+        Map<String, Set<String>> modelIdAndNodeIds = new HashMap<>();
+        queryModelStep.whenComplete(searchResponse -> {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            if (CollectionUtils.isEmpty(hits)) {
+                // no model need to be deployed
+                log.info("no model need to be deployed or undeployed");
+                return;
+            }
+
+            for (SearchHit hit : hits) {
+                try (XContentParser parser = MLNodeUtils.createXContentParserFromRegistry(xContentRegistry, hit.getSourceRef())) {
+                    ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                    MLModel mlModel = MLModel.parse(parser);
+                    String[] planningWorkerNodes = mlModel.getPlanningWorkerNodes();
+                    Set<String> nodeIds = Arrays.stream(planningWorkerNodes).collect(Collectors.toSet());
+
+                    if (isMasterNode) {
+                        if (allowCustomDeploymentPlan && !mlModel.isDeployToAllNodes()) {
```

Comment:
When allowCustomDeploymentPlan is false, the load API do not accept nodeIds as parameter, we shouldn't pass any nodeIds in this case.

Reply:
that's right~ When the `allowCustomDeploymentPlan` is false,whatever the `onlyRunOnMlNode` is true or false,the `nodeids` paramerter will be the empty collection, I will changed it.

Reply:
changed it,and I have committed it to this PR